pennylane/templates/subroutines/double_excitation_unitary.py

```diff
@@ -0,0 +1,550 @@
+    Hadamard(wires=q)
+    Hadamard(wires=p)
+
+
+def _layer7(weight, s, r, q, p, set_cnot_wires):
+    r"""Implement the seventh layer of the circuit to exponentiate the double-excitation
+    operator entering the UCCSD ansatz.
+
+    .. math::
+
+        \hat{U}_{pqrs}^{(7)}(\theta) = \mathrm{exp} \Big\{ -\frac{i\theta}{8}
+        \bigotimes_{b=s+1}^{r-1} \hat{Z}_b \bigotimes_{a=q+1}^{p-1} \hat{Z}_a
+        (\hat{Y}_s \hat{Y}_r \hat{Y}_q \hat{X}_p) \Big\}
+
+    Args:
+        weight (float): angle :math:`\theta` entering the Z rotation acting on wire ``p``
+        s (int): qubit index ``s``
+        r (int): qubit index ``r``
+        q (int): qubit index ``q``
+        p (int): qubit index ``p``
+        set_cnot_wires (sequence[int]): two-element sequence with the indices of the qubits
+            the CNOT gates act on
+    """
+
+    # U_1, U_2, U_3, U_4 acting on wires 's', 'r', 'q' and 'p'
+    RX(-np.pi / 2, wires=s)
+    RX(-np.pi / 2, wires=r)
+    RX(-np.pi / 2, wires=q)
+    Hadamard(wires=p)
+
+    # Applying CNOTs between wires 's' and 'p'
+    for cnot_wires in set_cnot_wires:
+        CNOT(wires=cnot_wires)
+
+    # Z rotation acting on wire 'p'
+    RZ(-weight / 8, wires=p)
+
+    # Applying CNOTs in reverse order
+    for cnot_wires in reversed(set_cnot_wires):
+        CNOT(wires=cnot_wires)
+
+    # U_1^+, U_2^+, U_3^+, U_4^+ acting on wires 's', 'r', 'q' and 'p'
+    RX(np.pi / 2, wires=s)
+    RX(np.pi / 2, wires=r)
+    RX(np.pi / 2, wires=q)
+    Hadamard(wires=p)
+
+
+def _layer8(weight, s, r, q, p, set_cnot_wires):
+    r"""Implement the eighth layer of the circuit to exponentiate the double-excitation
+    operator entering the UCCSD ansatz.
+
+    .. math::
+
+        \hat{U}_{pqrs}^{(8)}(\theta) = \mathrm{exp} \Big\{ -\frac{i\theta}{8}
+        \bigotimes_{b=s+1}^{r-1} \hat{Z}_b \bigotimes_{a=q+1}^{p-1} \hat{Z}_a
+        (\hat{Y}_s \hat{Y}_r \hat{X}_q \hat{Y}_p) \Big\}
+
+    Args:
+        weight (float): angle :math:`\theta` entering the Z rotation acting on wire ``p``
+        s (int): qubit index ``s``
+        r (int): qubit index ``r``
+        q (int): qubit index ``q``
+        p (int): qubit index ``p``
+        set_cnot_wires (sequence[int]): two-element sequence with the indices of the qubits
+            the CNOT gates act on
+    """
+
+    # U_1, U_2, U_3, U_4 acting on wires 's', 'r', 'q' and 'p'
+    RX(-np.pi / 2, wires=s)
+    RX(-np.pi / 2, wires=r)
+    Hadamard(wires=q)
+    RX(-np.pi / 2, wires=p)
+
+    # Applying CNOTs between wires 's' and 'p'
+    for cnot_wires in set_cnot_wires:
+        CNOT(wires=cnot_wires)
+
+    # Z rotation acting on wire 'p'
+    RZ(-weight / 8, wires=p)
+
+    # Applying CNOTs in reverse order
+    for cnot_wires in reversed(set_cnot_wires):
+        CNOT(wires=cnot_wires)
+
+    # U_1^+, U_2^+, U_3^+, U_4^+ acting on wires 's', 'r', 'q' and 'p'
+    RX(np.pi / 2, wires=s)
+    RX(np.pi / 2, wires=r)
+    Hadamard(wires=q)
+    RX(np.pi / 2, wires=p)
```

Comment:
Is there no numerical law that would allow you to implement one layer, but arrange it according to some hyperparameters that depend on a layer index passed to this function?

Reply:
Most probably, yes and it may be more elegant from a coding point of view. However, I am inclined to believe that in terms of readability and maintenance it is good to have a one-to-one correspondence with the  equations in the appendices of `arXiv:1805.04340 <https://arxiv.org/abs/1805.04340>`_.